File: cpp/src/builder/pose.cpp
/*
 * File: pose.cpp
 * Author: Minjun Xu (mjxu96@gmail.com)
 * File Created: Saturday, 7th December 2019 2:55:16 pm
 */

#include "xviz/builder/pose.h"

using namespace xviz;

XVIZPoseBuilder::XVIZPoseBuilder(const std::shared_ptr<xviz::Metadata>& metadata) : 
  XVIZBaseBuilder(xviz::StreamMetadata::POSE, metadata) {
  poses_ = nullptr;
  temp_poses_ = xviz::Pose();
}

void XVIZPoseBuilder::DeepCopyFrom(const XVIZPoseBuilder& other) {
  XVIZBaseBuilder::DeepCopyFrom(other);
  DeepCopyPtr(poses_, other.poses_);
  temp_poses_.CopyFrom(other.temp_poses_);
}

XVIZPoseBuilder& XVIZPoseBuilder::Stream(const std::string& stream_id) {
  if (stream_id_.size() > 0) {
    this->Flush();
  }
  stream_id_ = stream_id;
  return *this;
}

XVIZPoseBuilder& XVIZPoseBuilder::MapOrigin(double longitude, double latitude, double altitude) {
  temp_poses_.mutable_map_origin()->set_longitude(longitude);
  temp_poses_.mutable_map_origin()->set_latitude(latitude);
  temp_poses_.mutable_map_origin()->set_altitude(altitude);
  return *this;
}

XVIZPoseBuilder& XVIZPoseBuilder::Position(double x, double y, double z) {
  temp_poses_.add_position(x);
  temp_poses_.add_position(y);
  temp_poses_.add_position(z);
  return *this;
}

XVIZPoseBuilder& XVIZPoseBuilder::Orientation(double roll, double pitch, double yaw) {
  temp_poses_.add_orientation(roll);
  temp_poses_.add_orientation(pitch);
  temp_poses_.add_orientation(yaw);
  return *this;
}

XVIZPoseBuilder& XVIZPoseBuilder::Timestamp(double timestamp) {
  temp_poses_.set_timestamp(timestamp);
  return *this;
}

void XVIZPoseBuilder::Flush() {
  if (poses_ == nullptr) {
    poses_ = std::make_shared<std::unordered_map<std::string, xviz::Pose>>();
  }
  poses_->insert({stream_id_, temp_poses_});
  temp_poses_ = xviz::Pose();
}

std::shared_ptr<std::unordered_map<std::string, xviz::Pose>> XVIZPoseBuilder::GetData() {
  if (stream_id_.size() > 0) {
    Flush();
  }
  return poses_;
}

File: python/xviz_avs/builder/pose.py
from xviz_avs.builder.base_builder import XVIZBaseBuilder, CATEGORY
from xviz_avs.v2.core_pb2 import Pose, MapOrigin

class XVIZPoseBuilder(XVIZBaseBuilder):
    """
    # Reference
    [@xviz/builder/xviz-pose-builder]/(https://github.com/uber/xviz/blob/master/modules/builder/src/builders/xviz-pose-builder.js)
    """
    def __init__(self, metadata, logger=None):
        super().__init__(CATEGORY.POSE, metadata, logger)

        self._poses = None
        self.reset()

    def reset(self):
        super().reset()

        self._category = CATEGORY.POSE
        self._temp_pose = Pose()

    def map_origin(self, longitude, latitude, altitude):
        self._temp_pose.map_origin.longitude = longitude
        self._temp_pose.map_origin.latitude = latitude
        self._temp_pose.map_origin.altitude = altitude
        return self

    def position(self, x, y, z):
        self._temp_pose.position.extend([x, y, z])
        return self

    def orientation(self, roll, pitch, yaw):
        self._temp_pose.orientation.extend([roll, pitch, yaw])
        return self

    def timestamp(self, timestamp):
        self._temp_pose.timestamp = timestamp
        return self

    def _flush(self):
        if not self._poses:
            self._poses = {}

        self._poses[self._stream_id] = self._temp_pose
        self._temp_pose = Pose()

    def get_data(self):
        if self._stream_id:
            self._flush()

        return self._poses
